pkg/cover/report_test.go

```diff
@@ -152,7 +152,7 @@ void __sanitizer_cov_trace_pc() { printf("%llu", (long long)__builtin_return_add
 `)); err != nil {
 		t.Fatal(err)
 	}
-	kcovFlags := append([]string{"-c", "-w", "-x", "c", "-o", kcovObj, kcovSrc}, target.CFlags...)
+	kcovFlags := append([]string{"-c", "-w", "-x", "c", "-o", kcovObj, kcovSrc}, target.GenerateCFlags()...)
```

Comment:
Is it the only place that fails with ASLR? I am thinking maybe we could improve the test to not fail with ASLR enabled. It's pity to add this complexity throughout because of the way a single test is written.

Reply:
At least judging by our tests, yes. 
[This simpler implementation](https://github.com/a-nogikh/syzkaller/commit/e6a554ba0decfcb47d9100744c022678142b1ba6) triggers these errors:
```
--- FAIL: TestReportGenerator (6.81s)
    --- FAIL: TestReportGenerator/linux-amd64 (0.01s)
        --- FAIL: TestReportGenerator/linux-amd64/good-pie (0.48s)
            report_test.go:207: using exact coverage PC 0x55be70752004
            report_test.go:132: expected no error, but got:
                coverage doesn't match any coverage callbacks
        --- FAIL: TestReportGenerator/linux-amd64/good-pie-relocs (0.73s)
            report_test.go:207: using exact coverage PC 0x562659185004
            report_test.go:132: expected no error, but got:
                coverage doesn't match any coverage callbacks
        --- FAIL: TestReportGenerator/linux-amd64/good (0.88s)
            report_test.go:207: using exact coverage PC 0x7f903f7d4814
            report_test.go:132: expected no error, but got:
                coverage doesn't match any coverage callbacks
    --- FAIL: TestReportGenerator/linux-386 (0.01s)
        --- FAIL: TestReportGenerator/linux-386/good-pie (0.77s)
            report_test.go:207: using exact coverage PC 0xffffffff89873012
            report_test.go:132: expected no error, but got:
                coverage doesn't match any coverage callbacks
        --- FAIL: TestReportGenerator/linux-386/good (0.90s)
            report_test.go:207: using exact coverage PC 0xfffffffff7ee9892
            report_test.go:132: expected no error, but got:
                coverage doesn't match any coverage callbacks
FAIL
FAIL    github.com/google/syzkaller/pkg/cover   8.891s
```

Reply:
This fixes pkg/cover tests:

```
        if err := osutil.WriteFile(kcovSrc, []byte(`
+#define _GNU_SOURCE
+#include <dlfcn.h>
+#include <link.h>
+#include <stddef.h>
+
+void* aslr_base() {
+       void* handle = dlopen(NULL, RTLD_LAZY | RTLD_NOLOAD);
+       struct link_map* map = NULL;
+       dlinfo(handle, RTLD_DI_LINKMAP, &map);
+       return map ? map->l_addr : NULL;
+}
+
 #include <stdio.h>
-void __sanitizer_cov_trace_pc() { printf("%llu", (long long)__builtin_return_address(0)); }
+void __sanitizer_cov_trace_pc() { printf("%llu", (long long)(__builtin_return_address(0) - aslr_base())); }
 `)); err != nil {
```

dlinfo seems to be supported on FreeBSD as well:
https://www.freebsd.org/cgi/man.cgi?query=dlinfo&sektion=3&apropos=0&manpath=FreeBSD+13.0-RELEASE+and+Ports

I would not touch Akaros, it has very old toolchain and is not particularly interesting. I would also not touch any OSes if we can't easily fix pkg/cover test for them (e.g. missing dlinfo).